### PR TITLE
feat: start the service through configuration

### PR DESCRIPTION
**Pull request by Kree0**
# Required checklist
- [ ] Sample config files updated (`config`,`meta/config` and `default config`) 
- [ ] If there are user-facing changes, the documentation needs to be updated prior to approving the PR( [Link]() )
- [ ] If there are any breaking changes to public APIs, please add the `api change` label.
- [ ] Signed [CLA](https://cla-assistant.io/cnosdb/cnosdb) (if not already signed)

# Which issue does this PR close?

[//]: # (We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For -- example `Closes #123` indicates that this PR will close issue #123.)

Related #.

# Rationale for this change

[//]: # (Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.)

# Are there any user-facing changes?

[//]: # (There is no need to duplicate the description in the issue here, but it is sometimes worth providing a summary of the individual changes in this PR.)

### CODE REVIEW

**Comment by zipper-meng** [comment]
Write something to issue conversations to describe what you've changed to the configurations.

File: meta/src/model/meta_admin.rs [hunk 1]
```diff
@@ -414,11 +414,11 @@ impl AdminMeta {
             id: self.config.node_basic.node_id,
             grpc_addr: build_address(
                 self.config.host.clone(),
-                self.config.cluster.grpc_listen_port,
+                self.config.cluster.grpc_listen_port.unwrap(),
```
Comment (zipper-meng) [inline meta/src/model/meta_admin.rs:417]: Config `grpc_listen_port` may be none but here trying to give `grpc_addr` an exact value.

File: config/src/cluster_config.rs [hunk 2]
```diff
@@ -34,20 +34,20 @@ impl ClusterConfig {
         vec!["127.0.0.1:8901".to_string()]
     }
 
-    fn default_http_listen_port() -> u16 {
-        8902
+    fn default_http_listen_port() -> Option<u16> {
```
Comment (zipper-meng) [inline config/src/cluster_config.rs:37]: As you removed serde derive **default**, these functions `default_xxx_listen_port()` are useless. If you want to keep these default listened port in code, I have an idea but the config pattern needs to be modified:

```toml
[cluster]

[cluster.http]
enabled = true
listen_port = 8902

[cluster.grpc]
enabled = true
listen_port = 8903


```

File: common/models/src/utils.rs [hunk 3]
```diff
@@ -99,6 +99,9 @@ impl SeqIdGenerator {
     }
 }
 
-pub fn build_address(address: String, port: u16) -> String {
-    format!("{}:{}", address, port)
+pub fn build_address(address: String, port: Option<u16>) -> String {
```
Comment (zipper-meng) [inline common/models/src/utils.rs:102]: For parameter `address` I think we don't need to take an owned String. And, build an empty string as the address is dangerous. Try this:

```rust
pub fn build_address(address: &str, port: u16) -> String {
    format!("{address}:{port}")
}
```

**Review by zipper-meng — APPROVED** [review]
That's good

File: main/src/server.rs [hunk 4]
```diff
@@ -128,14 +128,15 @@ async fn regular_report_node_metrics(meta: MetaRef, heartbeat_interval: u64) {
...
-    build_address(DEFAULT_NODE_IP.to_owned(), port)
+    build_address(DEFAULT_NODE_IP, port)
 }
 
 impl ServiceBuilder {
     pub async fn build_storage_server(&self, server: &mut Server) -> Option<EngineRef> {
         let meta = self.create_meta().await;
-
-        meta.add_data_node().await.unwrap();
+        if meta.enabled() {
```
Comment (zipper-meng) [inline main/src/server.rs:137]: `add_data_node()` must be called whether a data node defined it's `grpc_addr`.

File: main/src/server.rs [hunk 5]
```diff
@@ -328,24 +361,26 @@ impl ServiceBuilder {
...
-    fn create_vector_grpc(
+    fn create_vector_grpc_if_enabled(
         &self,
         coord: CoordinatorRef,
         dbms: DBMSRef,
-        port: u16,
-    ) -> VectorGrpcService {
-        let default_vector_grpc_addr = build_default_address(port);
+    ) -> Option<VectorGrpcService> {
+        let default_vector_grpc_addr = match self.config.cluster.vector_listen_port {
```
Comment (bartliu827) [inline main/src/server.rs:380]: Is it better to put this judgment outside

Comment (zipper-meng) [reply main/src/server.rs:378]: Is this pattern better?

```rust
fn create_vector_grpc(
    &self,
    coord: CoordinatorRef,
    dbms: DBMSRef,
    port: u64,
) -> VectorGrpcService {
   ...
}

if let Some(port) = config.cluster.vector_listen_port {
    let vector_service = create_vector_grpc(...);
    server.add_service(Box::new(vector_service));
}
```

File: main/src/server.rs [hunk 6]
```diff
@@ -148,19 +147,26 @@ impl ServiceBuilder {
...
-        let tcp_service = Box::new(self.create_tcp(coord.clone()));
 
-        server.add_service(http_service);
-        server.add_service(grpc_service);
-        server.add_service(tcp_service);
+        if let Some(tcp_service) = self.create_tcp_if_enabled(coord.clone()) {
+            server.add_service(Box::new(tcp_service));
+        }
+
+        if let Some(vector_service) =
```
Comment (zipper-meng) [inline main/src/server.rs:165]: Here do not need to start a service of vector-grpc.

**Review by zipper-meng — APPROVED** [review]
It's ok.

**Pull request merged**